File: frameworks/CSharp/aspnetcore/Benchmarks/Middleware/PlaintextMiddleware.cs
// Copyright (c) .NET Foundation. All rights reserved. 
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 

using System;
using System.Text;
using System.Threading.Tasks;
using Benchmarks.Configuration;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;

namespace Benchmarks.Middleware
{
    public class PlaintextMiddleware
    {
        private static readonly PathString _path = new PathString(Scenarios.GetPath(s => s.Plaintext));
        private static readonly byte[] _helloWorldPayload = Encoding.UTF8.GetBytes("Hello, World!");

        private readonly RequestDelegate _next;
        
        public PlaintextMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public Task Invoke(HttpContext httpContext)
        {
            if (httpContext.Request.Path.StartsWithSegments(_path, StringComparison.Ordinal))
            {
                httpContext.Response.StatusCode = 200;
                httpContext.Response.ContentType = "text/plain";
                // HACK: Setting the Content-Length header manually avoids the cost of serializing the int to a string.
                //       This is instead of: httpContext.Response.ContentLength = _helloWorldPayload.Length;
                httpContext.Response.Headers["Content-Length"] = "13";
                return httpContext.Response.Body.WriteAsync(_helloWorldPayload, 0, _helloWorldPayload.Length);
            }

            return _next(httpContext);
        }
    }
    
    public static class PlaintextMiddlewareExtensions
    {
        public static IApplicationBuilder UsePlainText(this IApplicationBuilder builder)
        {
            return builder.UseMiddleware<PlaintextMiddleware>();
        }
    }
}

File: frameworks/VB/aspnetcore/Benchmarks/Middleware/PlaintextMiddleware.vb
' Copyright (c) .NET Foundation. All rights reserved. 
' Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 

Imports System.Runtime.CompilerServices
Imports System.Text
Imports Microsoft.AspNetCore.Builder
Imports Microsoft.AspNetCore.Http

Public Class PlaintextMiddleware

    Private Shared ReadOnly HelloWorldPayload As Byte() = Encoding.UTF8.GetBytes("Hello, World!")
    Private ReadOnly NextStage As RequestDelegate

    Public Sub New(ByVal NextStage As RequestDelegate)
        Me.NextStage = NextStage
    End Sub

    Public Function Invoke(ByVal httpContext As HttpContext) As Task

        If httpContext.Request.Path.StartsWithSegments("/plaintext", StringComparison.Ordinal) Then
            Return WriteResponse(httpContext.Response)
        End If

        Return NextStage(httpContext)

    End Function

    Public Shared Function WriteResponse(ByVal response As HttpResponse) As Task

        Dim payloadLength = HelloWorldPayload.Length
        response.StatusCode = 200
        response.ContentType = "text/plain"
        response.ContentLength = payloadLength
        Return response.Body.WriteAsync(HelloWorldPayload, 0, payloadLength)

    End Function

End Class

Module PlaintextMiddlewareExtensions
    <Extension()>
    Function UsePlainText(ByVal builder As IApplicationBuilder) As IApplicationBuilder
        Return builder.UseMiddleware(Of PlaintextMiddleware)()
    End Function
End Module
